library/sonar_plugin.py

```diff
@@ -0,0 +1,429 @@
+            current_key = re.sub('-', '', plugin)
+            if current_key == key:
+                ready_data = re.sub(r'.*sonar', 'sonar', data)
+                final_url = repo_url + data.lstrip() + re.sub('/', '', ready_data) + '-' + version + '.jar'
+                return final_url
+
+    return final_url
+
+
+def is_plugin_update_available(module, sonar_url, updates_list_api, installed_list_api):
+    is_available = False
+    version_installed = 'not-found'
+    url = sonar_url + installed_list_api
+    response = requests.get(url, auth=HTTPBasicAuth(module.params['username'],
+                                                    module.params['password']))
+    json_obj = response.json()
+
+    for plugin in json_obj['plugins']:
+        if plugin['name'] == module.params['name']:
+            version_installed = plugin['version']
+
+    comparison = compare_plugins_version(module.params['version'],
+                                         version_installed)
+
+    if comparison == 'Equal':
+        module.exit_json(changed=False, stdout='FAILED: Version already installed')
+    elif comparison == 'Lower':
+        is_available = True
+    else:
+        url = sonar_url + updates_list_api
+        response = requests.get(url, auth=HTTPBasicAuth(module.params['username'], module.params['password']))
+        json_obj = response.json()
+        for plugin in json_obj['plugins']:
+            if plugin['name'] == module.params['name']:
+                for update in plugin['updates']:
+                    if compare_plugins_version(module.params['version'], update['release']['version']) == 'Equal':
+                        if not update['requires'] and update['status'] == 'COMPATIBLE':
+                            is_available = True
+
+    return is_available
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str'),
+            state=dict(type='str', choices=['installed', 'removed', 'latest']),
+            version=dict(type='str'),
+            custom_url=dict(type='str'),
+            hostname=dict(default='localhost', type='str'),
+            username=dict(default='admin', type='str'),
+            password=dict(default='admin', type='str', no_log=True),
+            pending_dir=dict(default='/usr/local/sonar/extensions/downloads/', type='str'),
+            sonar_port=dict(default='9000', type='int')
+        ),
+        mutually_exclusive=[['custom_url', 'state']]
+    )
+
+    available_list_api = '/api/plugins/available'
+    installed_list_api = '/api/plugins/installed'
+    updates_list_api = '/api/plugins/updates'
+
+    install_api = '/api/plugins/install'
+    remove_api = '/api/plugins/uninstall'
+    update_api = '/api/plugins/update'
+
+    sonarsource_plugins_repo = 'https://binaries.sonarsource.com/Distribution/'
+
+    state = module.params['state']
+    sonar_url = 'http://' + module.params['hostname'] + ':' + str(module.params['sonar_port'])
```

Comment:
you have better ways to format strings in Python
and if you already created a 'state' var then do it for others too.

Reply:
+1 @abraverm 
Try using format() method for this purpose, looks much more understandable.

Reply:
done